### DISCUSSION TITLE
Combining `uses_from_macos` with `on_macos`?

### DISCUSSION THREAD

**Discussion by gwerbin** [Homebrew/discussions#1766] [Writing Formulae/Casks] [answered]
Hello all,

I am working on the [Idris 2 formula](https://github.com/Homebrew/homebrew-core/blob/ee2212664e8c609d8186cfabd3feac407f028982/Formula/idris2.rb), which currently specifies:
```ruby
uses_from_macos "zsh" => :build, since: :mojave
```

This is not strictly correct. The actual logic should be something like this:

* If on Mac, use the system-provided Zsh when available (starting with Mojave), otherwise use Homebrew Zsh.
* If on Linux, don't use Zsh at all, for anything.

That is, Zsh is _only a build dependency on Mac_, and is not required at all on Linux.

I tried to write:
```ruby
if on_macos
  uses_from_macos "zsh" => :build, since: :mojave
end
```

But this is apparently not allowed:

```
Error: 1 problem in 1 formula detected
idris2:
  * 19: col 5: `on_macos` cannot include `uses_from_macos`. Only `depends_on`, `patch`, `resource`, `deprecate!`, `disable!`, `conflicts_w
ith`, and `keg_only` are allowed.
```

Is there some workaround to enable Zsh-from-Macos as a build dependency on Mac systems, without forcing Linux users to have Zsh when it's entirely unnecessary for them? Is there any technical reason why this shouldn't be allowed?

**Comment by gromgit** [comment]
> If on Mac, use the system-provided Zsh when available (starting with Mojave), otherwise use Homebrew Zsh.

Why _must_ you use system Zsh on supported macOS? If you relax that restriction, the solution is simple:
```
on_macos do
  depends_on "zsh" => :build
end
```
If you don't intend to support macOS before Mojave, it's even simpler: Drop the dependency entirely, and Homebrew will automatically use system Zsh on all platforms.

**Reply by gwerbin** [reply]
I see no reason to require an external dependency when a perfectly good one is already available on the user's system.

Likewise, I see no reason to arbitrarily drop support for a platform, because the package manager just so happens to not support that specific combination of features.

**Reply by carlocab** [reply]
You can also do
```ruby
on_macos do
  depends_on "zsh" => :build if MacOS.version < :mojave
end
```

**Reply by gwerbin** [reply]
@carlocab that makes sense! Is that pretty much equivalent to what `uses_from_macos` does?

**Reply by carlocab** [reply]
Yes.
```ruby
uses_from_macos "zsh", since: :mojave
```
is a no-op on macOS Mojave or newer, and becomes
```ruby
depends_on "zsh"
```
on Linux and macOS High Sierra or older.

**Comment by gwerbin** [comment, marked as answer]
As per @carlocab , the workaround is:

```ruby
on_macos do
  depends_on "zsh" => :build if MacOS.version < :mojave
end
```
